Shared metrics context (Client -> SDK, SDK -> Client)

UWTech opened the pull request:
Add shared context for metrics between Client and SDK

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
+  * @param {string} eventName name of the metric
+  * @param {string} id unique identifier
+  * @param {Object} props to populate the metric with
+  * @returns {bool} true if metric registered, false otherwise
```
Is this bool expected to provide value outside of tests? If not, a spy may be more appropriate.

UWTech replied:
The intention here is for the bool to potentially be used by whatever consumer is requesting a metric be registered. This could be used to determine if there is some problem with the metric registration/deletion process since this will be managed by way of register/update/send methods.

In the event a metric has already been registered with that ID, the consumer could read the FALSE return, and take some action (console log, send failure metric, secondary hashing, etc.).

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
+    }
+
+    // update metric
+    _.assign(props.fields, metricProps.fields);
```
Use `Object.assign`. Generally, you want to avoid using lodash methods if there's a native method available.

UWTech replied:
done

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -55,9 +67,12 @@ const Metrics = SparkPlugin.extend({
+    return this.sendBatchMetric(payload);
   },
 
+  sendBatchMetric(payload) {
```
This should probably be prefixed with an `_`.

UWTech replied:
done

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -26,14 +32,7 @@ const Metrics = SparkPlugin.extend({
-   * @returns {Object} HttpResponse object
-   */
-  submitClientMetrics(eventName, props, preLoginId) {
+  createPayload(eventName, props) {
```
Please add doc block (to all new methods).

UWTech replied:
done

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
+  * @param {Object} props to populate the metric with
+  * @returns {bool} true if metric registered, false otherwise
+  */
+  clientRegisterMetric(eventName, id, props) {
```
Let's call this `registerClientMetric()`.

UWTech replied:
done

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
+  * @param {Object} props to populate the metric with
+  * @returns {bool} true if metric registered, false otherwise
+  */
+  sdkRegisterMetric(eventName, id, props) {
```
Let's call this `registerMetric()`. 

If there's already a client metric, this should behave like `updateRegisteredMetric()`, right?

UWTech replied:
This is an abstraction of the underlying implementation so we can change the functionality of the method later. It is not strictly necessary, but prevents us from having to use the same preprocessing for both client and SDK metrics.

Is just a simple wrapper.

ianwremmel replied:
yea, understood. i was thinking it made more sense for the method called by the sdk to be a pretty generic name, and then the method called by consumers would be obviously for consumers.

UWTech replied:
Right, right; got it. Since we're thinking about this in the context of the SDK, the generic method should be used by the SDK, whereas the explicitly differentiated one would be for external users.

I can change the names of the SDK ones, sure.

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -26,14 +32,7 @@ const Metrics = SparkPlugin.extend({
-   * @returns {Object} HttpResponse object
-   */
-  submitClientMetrics(eventName, props, preLoginId) {
+  createPayload(eventName, props) {
```
Done

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
+    return this.metricMap.delete(id);
+  },
+
+  registerMetric(eventName, id, props, type) {
```
done

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
+   * @param {string} id of the metric to be sent
+   * @returns {Object} HttpResponse object
+   */
+  clientSendRegisteredMetric(id) {
```
So this method actually sends the metric. Register simply creates and stores the metric

ianwremmel replied:
gotcha, yea, i misread the name a bit. similarly to the sdk method above, it seems to me the use-by-sdk methods should be generically named and the use-by-client methods should have the "client" moniker. that said, `sendClientRegisteredMetric()` would be more appropriate (method names should generally be verbs or, at least, verb-like).

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
+   * @param {string} id of the metric to be sent
+   * @returns {Object} HttpResponse object
+   */
+  sdkSendRegisteredMetric(id) {
```
So the idea here is that if you are the original creator of the metric, be you the client or the sdk, you are the party that has ultimate authority to send the metric. In order to otherwise know the role of the party, we would have to pass a param to the method.

Using a wrapper like this ensures that we are not unintentionally sending a metric that is registered and owned by the client in the SDK, or vice versa. 

I'm not sure I understand the argument for only having a single send metric method. Are you suggesting a paramaterized version where we send the identifier as an argument, (e.g.: "SDK" or "Client")?

So that would look like sendMetric(id, requestingParty){
if(getMetric(id).type != requestingParty) {
// party does not own this metric, cannot send, only update!
return false
}
else {
send metric
}

ianwremmel replied:
i was still thinking there'd be two send methods: `sendRegisteredMetric()` and `sendRegisteredClientMetric()`. `sendRegisteredMetric()` should be a noop if `registerClientMetric()` is called for that key before `registerMetric()` is called for that key. That way, the sdk can report a given dataset without client input via 

- sdk: `registerMetric()`
- sdk: `updateMetric()` 
- sdk: `sendMetric()`

should a client choose to participate, the flow would be 

- client: `registerClientMetric()`
- sdk: `registerMetric()`
- sdk: `updateMetric()` 
- sdk: `sendMetric()` (noop in this case)
- client: `sendClientMetric()`

UWTech replied:
done

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/test/unit/spec/metrics.js:
```diff
@@ -172,5 +173,150 @@ describe(`plugin-metrics`, () => {
+
+    describe(`#sendUnstructured`, () => {
+      it(`sends an unstructured metric`, () => {
+        metrics.sendUnstructured(`foo`, `bar`)
```
This "test" was just filler. I was just trying to generate coverage because the pre-check was failing. I was only going for coverage rather than value. Figured that was okay since the method was deprecated and previously uncovered by tests anyway.

I'll double check the others to be sure though.

ianwremmel replied:
yea, but that would mean that, should this test start failing at some point, we'd never know.

UWTech replied:
fixed

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -7,6 +7,7 @@ import Batcher from './batcher';
 import ClientMetricsBatcher from './client-metrics-batcher';
 import CallDiagnosticEventsBatcher from './call-diagnostic-events-batcher';
 import {deprecated} from '@ciscospark/common';
+import _ from 'lodash';
```
Removed all together. Using Object.assign instead.

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -15,6 +16,11 @@ const Metrics = SparkPlugin.extend({
     callDiagnosticEventsBatcher: CallDiagnosticEventsBatcher
   },
 
+  metricMap: new Map(),
```
Wouldn't we want to share the context information across instances?

I could conceive of situations where we'd want to share the information across different modules that may not necessarily be talking to the same instance.

I believe in the case of the web-client, we do have a single instance of Spark-JS-SDK, but seems it might be useful.

Is the thinking this is not expected behavior from most people; e.g.: if I create a new instance, I expect all the context information to be fresh, and that the instance does not consider the other metrics I've started to create for the purposes of sharing between the SDK and the client?

ianwremmel replied:
the metrics plugin will be instantiated once for each sdk instance. Typically, you would only make a second sdk instance if you had a different set of credentials (e.g. for a node web server that handles multiple user requests). In those cases, you definitely don't want to share the metrics map across instances.

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
+    return this.metricMap.set(id, tuple);
+  },
+
+  getMetrics() {
```
done

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/test/unit/spec/metrics.js:
```diff
@@ -172,5 +173,150 @@ describe(`plugin-metrics`, () => {
+      it(`sends client metric`, () => {
+        metrics.registerMetric(`horse`, `1`, testPayload, `clientMetric`);
+        metrics.clientSendRegisteredMetric(`1`)
+        .then(() => {
```
done

UWTech commented:
Addressed PR comments.

Not super familiar with Ampersand, so let me know if there is some other pattern you'd like followed for the map instantiation.

Cheers.

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/test/unit/spec/metrics.js:
```diff
@@ -172,5 +173,150 @@ describe(`plugin-metrics`, () => {
+            assert.equal(metric.fields.perceivedDurationInMillis, testPayload.fields.perceivedDurationInMillis);
+          });
+      });
+      it(`does not send non-sdj metric`, () => {
```
"sdk"

UWTech replied:
done

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -99,6 +121,111 @@ const Metrics = SparkPlugin.extend({
+  },
+
+  metricIsType(id, type) {
+    return this.getMetric(id) && this.getMetric(id).type === type;
```
store the result of `getMetric()` so that you're only doing one lookup. you could also use lodash's `get()`

```javascript
return get(this.getMetric(id), `type`) === type;
```

UWTech replied:
done
Not sure there is a big cost differential between to two O(1) lookups versus comparing local variable, and one uses more bytes, but is a small enough change.

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/test/unit/spec/metrics.js:
```diff
@@ -172,5 +173,150 @@ describe(`plugin-metrics`, () => {
+    describe(`#sendUnstructured`, () => {
+      it(`sends an unstructured metric`, () => {
+        return metrics.sendUnstructured(`foo`, `bar`)
+        .then(() => {
```
indent `.then()`

UWTech replied:
done

ianwremmel commented on packages/node_modules/@ciscospark/internal-plugin-metrics/test/unit/spec/metrics.js:
```diff
@@ -172,5 +173,150 @@ describe(`plugin-metrics`, () => {
+
+    describe(`#sendUnstructured`, () => {
+      it(`sends an unstructured metric`, () => {
+        return metrics.sendUnstructured(`foo`, `bar`)
```
remove the return, move `metrics...` up a line, and use the arrow function's implicit return.

UWTech replied:
done

UWTech commented on packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js:
```diff
@@ -15,6 +15,13 @@ const Metrics = SparkPlugin.extend({
   },
 
+  props: {
+    metricMap: Map
```
done

UWTech commented:
Made changes as requested.

UWTech commented:
Created README

UWTech commented:
@ianwremmel  any word on when we might get this merged in?

No specific task we have right now that requires this, but would be nice to get this in for use around metrics we may need.

Cheers.
